File: frontend/src/api/posts/comments.js
import { checkSessionResponse } from '@/helpers/auth/auth'
import { router } from '@/router/router'

async function requestComments(url, options = {}) {
  const response = await fetch(url, {
    credentials: 'include',
    ...options,
  })

  if (!checkSessionResponse(response)) {
    router.replace('/login')
    return
  }

  const result = await response.json()
  if (!response.ok) {
    throw new Error(result.message || 'Could not load comments')
  }

  return result
}

export async function getComments(postId, { limit = 20, offset = 0 } = {}) {
  const params = new URLSearchParams({
    limit: String(limit),
    offset: String(offset),
  })

  return requestComments(`/api/posts/${postId}/comments?${params}`)
}

export async function createComment(postId, comment) {
  return requestComments(`/api/posts/${postId}/comments`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ content: comment.content }),
  })
}

export async function deleteComment(postId, commentId) {
  return requestComments(`/api/posts/${postId}/comments/${commentId}`, {
    method: 'DELETE',
  })
}

File: database/posts/comments.go
package posts

import (
	"database/sql"
	"errors"
)

var ErrCommentNotFound = errors.New("comment not found")

func DeleteComment(db *sql.DB, userID int, postID, commentID int64) error {
	result, err := db.Exec(
		`DELETE FROM comments WHERE id = ? AND post_id = ? AND user_id = ?`,
		commentID, postID, userID,
	)
	if err != nil {
		return err
	}

	affected, err := result.RowsAffected()
	if err != nil {
		return err
	}
	if affected == 0 {
		return ErrCommentNotFound
	}

	return nil
}
